keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1067 @@
+  Rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language)
+  and
+  [AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+- Version conversion support (similar to webhooks MatchPolicy)
+- Access the old object (similar to [transition
+  rules](https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/2876-crd-validation-expression-language#transition-rules)
+  and oldObject in AdmissionRequest)
+- Configurability, as motivated above.
+
+There are also lots of details (response message formatting, severity
+levels, failure policies, type safety, ...) that will be discussed in detail
+further on in this proposal.
+
+#### Configurability in the API
+
+How configurability is expressed in the API will have significant influence
+how the API is shaped. We are currently considering two design options:
+
+##### Configurability Option 1: Embed configuration into the object, grouped with match criteria
+
+```
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionConfiguration
+metadata:
+  name: "validate-xyz.example.com"
+configurationSchema:  
+  openAPIV3Schema: ... # the below configuration is validated against this schema
+spec:
+  match:
+  - rules: ... # see ValidatatingWebhookConfiguration rules
+    objectSelectors: ...  # "
+    namespaceSelectors ... # "
+    configuration: {x: 1} # each match may be configured differently
+  - rules: ...
+    objectSelectors: ...
+    namespaceSelectors ...
+    configuration: {x: 2}
+validations:
+  - rule: "self.spec.xyz == configuration.x"
+    # ...other rule related fields here...
+```
+
+Pros:
+
+- Self-contained resource
+- Higher level policy abstractions can be defined out-of-tree and can be
+  translated into this form by controllers
+
+Cons:
+
+- The resource grows by the number of match rules and amounts of configuration data.
+- Need to define how configurations of overlapping matches are interpreted. If
+  any of the matches fail does the rule as an admission check fail, or if any
+  matches pass does the admission check as a whole pass?
+
+##### Configurability Option 2: Put configuration in separate resources
+
+```
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionConfiguration
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  configurationCustomResourceDefinition:
+    group: rules.example.com
+    kind: LimitRule
+    ...
+      openAPIV3Schema: ... # the below configuration is validated against this schema
+  validations:
+    - rule: "self.spec.xyz == configuration.x"
+      # ...other rule related fields here...
+    ...
+---
+apiVersion: rules.example.com/v1
+kind: LimitRule
+metadata:
+  name: "validate-xyz-rule1.example.com"
+spec:
+  match:
+  - rules: ...
+    objectSelectors: …
+    namespaceSelectors ... 
```

Comment:
was the match/rules portion supposed to be included in this custom resource, or just the `x: 1` bit?

if so, separating the validation rules from the match rules seems like it would make it *really* hard to type check and understand the safety of changes to either object

Reply:
Matching and configuration both go into the custom resource.

This allows policy to be configured differently for different resources.  I've added more examples of this in the updated KEP.

re: type safety.  Yes. My current thinking is that the policy definition (`ValidatingAdmissionPolicy`) needs to establish a constraint on which GVKs it can validated so that type-checking can be performed and then all the custom resources that configure the policy need to be constrained to that GVK. I'm working on clearing this up in the KEP.